src/charm.py

```diff
@@ -0,0 +1,477 @@
+            event.add_status(ops.WaitingStatus("Waiting for ingress."))
+
+    @staticmethod
+    def _generate_secret_key(length=128):
+        """Return a random sequence of characters suitable for a secret key."""
+        choice = random.SystemRandom().choice
+        return "".join(choice(string.ascii_uppercase + string.digits) for _ in range(length))
+
+    def _on_install(self, event: ops.InstallEvent):
+        if not self.unit.is_leader():
+            logger.debug("Non-leader doesn't need to handle install.")
+            return
+        if not self.peers:
+            logger.debug("Waiting for peer relation before creating secret key.")
+            return
+        self._add_secret_key()
+
+    def _on_peer_relation_created(self, event: ops.RelationCreatedEvent):
+        if not self.unit.is_leader():
+            logger.debug("Non-leader doesn't need to handle peer-relation-created.")
+            return
+        self._add_secret_key()
+
+    def _add_secret_key(self):
+        """Generate a secret key for Takahē and store it as a Juju secret."""
+        if not self.peers:
+            logger.warning("Cannot add secret key before peer relation exists.")
+            return
+        # This is used for internal cryptography.
+        secret_key = self._generate_secret_key()
+        secret = self.unit.add_secret({"takahe-secret-key": secret_key})
+        if secret.id is None:
+            # TODO: Can this be tightened up in the ops type hinting?
+            logger.warning("Juju secret is missing ID")
+            return
+        self.peers.data[self.app]["secret-id"] = secret.id
+        logger.debug("Set Takahē secret key, id %s", secret.id)
+
+    def _on_cycle_secret_key(self, event: ops.ActionEvent):
+        if not self.unit.is_leader():
+            event.fail("Please run this action on the leader unit.")
+            return
+        if not self.peers:
+            event.fail("No need to cycle - we haven't even started up!")
+            return
+        try:
+            secret_id = self.peers.data[self.app]["secret-id"]
+        except KeyError:
+            event.fail("No need to cycle - have not created the initial secret yet!")
+            return
+        try:
+            secret = self.model.get_secret(id=secret_id)
+        except ops.SecretNotFoundError:
+            event.fail("No need to cycle - initial secret making in progress!")
+            return
+
+        new_secret_key = self._generate_secret_key()
+        secret.set_content({"takahe-secret-key": new_secret_key})
+        # The secret-changed event will take care of making use of the new revision.
+        event.set_results({"result": "success"})
+
+    def _on_add_superuser_action(self, event: ops.ActionEvent):
+        # This can be done on either container, so use the background service
+        # to minimise the impact.
+        container = self.unit.get_container(self.background_service_name)
+        env = self._takahē_env.copy()
```

Comment:
> Actually, looking at this now, I thought: if I was maintaining this charm code, how would I type this?

If on Ubuntu and using a Māori keyboard, then by holding down right-alt, press `e`. If on Ubuntu and not using a Māori keyboard then the compose key (I don't know what the default compose key is, though) then `-` then `e`. If on macOS it's option-e if I recall correctly. On iOS or iPadOS with the virtual keyboard by holding down `e` and picking `ē`. Windows has something similar to Ubuntu I think, but it's a long time since I've regularly used Windows.

> I think it's just going to cause pain when maintaining this to try to use `ē` in variables names etc in code. It means you can't type it without copying and pasting an `ē`, or using autocomplete. I know it's nice to be "correct", but maybe we should reserve the "ē" for just the product name in text/docs?

To be honest, this seems a bit Anglo-centric :smile:. But then I type non-Latin characters ("ngā mihi", "café", "Ionuţ", etc) pretty frequently and have for years, so maybe it's actually the opposite and it's my own bias :smile:.

I'd prefer to keep it for now, but will think on it :smile:.

Reply:
Fair push-back, and by all means keep it here ... it's a good way to test Python's Unicode handling if nothing else!

My take above is definitely Anglo-centric, I'll grant that. :-) But then again, open source code kinda is too, right? I agree we should try to get it right when we're using names (for people and things), I'm just not so sure it's a good idea in source code...

And TIL about the Compose Key -- thanks! I'd heard of that concept but thought it was a Mac-only thing. Then again, I spent about 20 minutes just now trying to figure out what the default is in Ubuntu (with a standard US keyboard layout). I've pressed it accidentally several times in the past (now that I understand what was going on), but can't repro that. I'll ask around.